rpc/client.go

```diff
@@ -91,219 +95,230 @@ type Client struct {
 //
 // The Client.Ready channel is closed after the client has connected
 // and completed one successful heartbeat. The Closed channel is
-// closed if the client fails to connect or if the client's Close()
-// method is invoked.
-func NewClient(addr net.Addr, opts *retry.Options, context *Context) *Client {
+// closed if the client's Close() method is invoked.
+func NewClient(addr net.Addr, context *Context) (*Client, error) {
 	clientMu.Lock()
+	defer clientMu.Unlock()
+
+	network := addr.Network()
+	address := addr.String()
+
+	key := network + address
+
 	if !context.DisableCache {
-		if c, ok := clients[addr.String()]; ok {
-			clientMu.Unlock()
-			return c
+		if c, ok := clients[key]; ok {
+			return c, nil
 		}
 	}
+
+	tlsConfig, err := context.GetClientTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
 	c := &Client{
-		addr:         addr,
-		Ready:        make(chan struct{}),
-		Closed:       make(chan struct{}),
-		clock:        context.localClock,
-		remoteClocks: context.RemoteClocks,
-		cached:       !context.DisableCache,
+		Closed:        make(chan struct{}),
+		key:           key,
+		network:       network,
+		remoteAddress: address,
+		tlsConfig:     tlsConfig,
+		clock:         context.localClock,
+		remoteClocks:  context.RemoteClocks,
 	}
-	if !context.DisableCache {
-		clients[c.Addr().String()] = c
+
+	c.healthy.Store(make(chan struct{}))
+
+	if err = c.connect(); err != nil {
+		return nil, err
 	}
-	clientMu.Unlock()
+
+	clients[key] = c
+
+	retryOpts := clientRetryOptions
+	retryOpts.Stopper = context.Stopper
 
 	context.Stopper.RunWorker(func() {
-		if err := c.connect(opts, context); err != nil {
-			log.Errorf("client %s failed to connect: %v", c.addr, err)
-		} else {
-			// Signal client is ready by closing Ready channel.
-			log.Infof("client %s connected", c.addr)
-			close(c.Ready)
+		<-context.Stopper.ShouldStop()
 
-			// Launch periodic heartbeat. This blocks until the client is to be closed.
-			c.runHeartbeat(context.Stopper)
-		}
 		c.Close()
 	})
-	return c
+
+	go func() {
```

Comment:
It's intentionally not a `RunWorker` because I wanted to avoid selecting on `c.Closed` **and** `stopper.ShouldStop()` everywhere; instead all the internals select on `c.Closed` and the `RunWorker` above waits on `ShouldStop()` and then calls `c.Close()`.